x/ibc/03-connection/types/version.go

```diff
@@ -1,96 +1,171 @@
 package types
 
-// GetCompatibleVersions returns an ordered set of compatible IBC versions for the
-// caller chain's connection end.
+import (
+	"fmt"
+	"strings"
+
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	host "github.com/cosmos/cosmos-sdk/x/ibc/24-host"
+)
+
+var (
+	// DefaultIBCVersion represents the latest supported version of IBC.
+	// The current version supports only ORDERED and UNORDERED channels.
+	DefaultIBCVersion = CreateVersionString("1", []string{"ORDERED channel", "UNORDERED channel"})
+)
+
+// GetCompatibleVersions returns a descending ordered set of compatible IBC
+// versions for the caller chain's connection end. The latest supported
+// version should be first element and the set should descend to the oldest
+// supported version.
 func GetCompatibleVersions() []string {
-	return []string{"1.0.0"}
+	return []string{DefaultIBCVersion}
 }
 
-// LatestVersion gets the latest version of a connection protocol
+// CreateVersionString constructs a valid connection version given a
+// version identifier and feature set. The format is written as:
 //
-// CONTRACT: version array MUST be already sorted.
-func LatestVersion(versions []string) string {
-	if len(versions) == 0 {
-		return ""
-	}
-	return versions[len(versions)-1]
+// ([version_identifier],[feature_0,feature_1,feature_2...])
+//
+// A connection version is considered invalid if it is not in this format
+// or violates one of these rules:
+// - the version identifier is empty or contains commas
+// - a specified feature contains commas
+func CreateVersionString(identifier string, featureSet []string) string {
+	return fmt.Sprintf("(%s,%s)", identifier, featureSet)
 }
 
-// PickVersion picks the counterparty latest version that is matches the list
-// of compatible versions for the connection.
-func PickVersion(counterpartyVersions, compatibleVersions []string) string {
+// UnpackVersion parses a version string and returns the identifier and the
+// feature set of a version. An error is returned if the version is not valid.
+func UnpackVersion(version string) (string, []string, error) {
+	// validate version so valid splitting assumptions can be made
+	if err := host.ConnectionVersionValidator(version); err != nil {
+		return "", nil, err
+	}
 
-	n := len(counterpartyVersions)
-	m := len(compatibleVersions)
+	// peel off prefix and suffix of the tuple
+	version = strings.TrimPrefix(version, "(")
+	version = strings.TrimSuffix(version, ")")
 
-	// aux hash maps to lookup already seen versions
-	counterpartyVerLookup := make(map[string]bool)
-	compatibleVerLookup := make(map[string]bool)
+	// split into identifier and feature set
+	splitVersion := strings.SplitN(version, ",", 2)
```

Comment:
ditto check for `splitVersion[0] == version`

Reply:
> Edge cases for s and sep (for example, empty strings) are handled as described in the documentation for Split.

Reply:
hmm I was working under the assumption the regex is correct. I don't think the edge cases could ever arise. Should I still add the checks?

Reply:
I think it's good to have the safe check but as you wish

Reply:
added the check :+1: